File: Week 2/id_545/LeetCode_590_545.js
// 590. N叉树的后序遍历

/*
// 递归
class Post {
    constructor() {
        this.arr = []
    }
    traversal(root) {
        if (!root) return []
        for(let i=0; i< root.children.length; i++) {
            this.traversal(root.children[i])
        }
        this.arr.push(root.val)
        return this.arr
    }
}
var postorder = function(root) {
    let p = new Post()
    return p.traversal(root)
};
*/

//栈 迭代
var postorder = function(root) {
    if (!root) return []
    let stack = [root]
    let arr = []
    while (stack.length != 0) {
        curr = stack.pop()
        arr.push(curr.val)
        for (let i in curr.children) {
            stack.push(curr.children[i])
        }
    }
    arr.reverse()
    return arr
};

File: Week 2/id_545/LeetCode_590_545.py
## 590. N叉树的后序遍历

# 递归
class Solution:
    def __init__(self):
        self.arr = []
    def postorder(self, root: 'Node') -> List[int]:
        if root:
            for i in range(len(root.children)):
                self.postorder(root.children[i])
            self.arr.append(root.val)
        return self.arr

# 栈 迭代法
# class Solution:
#     def postorder(self, root: 'Node') -> List[int]:
#         if not root: return []
#         stack = [root]
#         arr = []
#         while stack:
#             tmp = stack.pop()
#             arr.append(tmp.val)
#             if tmp.children:
#                 for i in tmp.children:
#                     stack.append(i)
#         return arr[::-1]